### PR TITLE
com.azure.core.http.HttpHeaders as public response headers contract

### PR DESCRIPTION
**Pull request by kirankumarkolli**
**Performance**: Current implementation has few draw-backs
- Maps are at-least coped twice on both request & response paths
- StoreResponse: GetHeaderValue (sequential N * M complexity) (LSN, PKRangeId, RequestCharges are always called by HA layer on store, that's the min overhead)
- Exception paths take one more copy of the headers 

**Changes**
- Response headers contract changed to com.azure.core.http.HttpHeaders
- Internal httpHeader types (nested consts groups) are renamed appropriately or removed
- Unnecessary tmp objects materialization avoided

With E2E leveraging the core.http.HttpHeaders the performance is kind of close. 
But future HttpHeaders optimizations should help get very nice gains.

**Future work**
- Work with azure.core team to further optimize the HttpHeaders implementation.

Fixes: #11222

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Utils.java [hunk 1]
```diff
@@ -68,6 +70,21 @@
         Utils.simpleObjectMapper.registerModule(new AfterburnerModule());
     }
 
+    public  static HttpHeaders clone(HttpHeaders inputHeaders) {
+        HttpHeaders httpHeaders = new HttpHeaders();
+        inputHeaders.forEach(entry -> httpHeaders.put(entry.getName(), entry.getValue()));
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Utils.java:75]: potential minor perf impact implication.

previously we used to pass the the old headers to the constructor of the new HashMap to allow reserving the required memory upfront. this way the underlying header map may grow multiple times.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/StoreClient.java [hunk 2]
```diff
@@ -132,24 +131,24 @@ private RxDocumentServiceResponse completeResponse(
             throw new InternalServerErrorException(RMResources.InvalidBackendResponse);
         }
 
-        Map<String, String> headers = new HashMap<>(storeResponse.getResponseHeaderNames().length);
+        HttpHeaders httpHeaders = new HttpHeaders();
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/StoreClient.java:134]: potential perf impact. previously we used to pass the capacity to the constructor of the hashmap upfront reserving the required capacity whereas now the underlying map can grow multiple times.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/StoreClient.java [hunk 3]
```diff
@@ -172,10 +171,10 @@ private void updateResponseHeader(RxDocumentServiceRequest request, Map<String,
             return;
         }
 
-        String partitionKeyRangeId = headers.get(WFConstants.BackendHeaders.PARTITION_KEY_RANGE_ID);
+        String partitionKeyRangeId = headers.getValue(WFConstants.BackendHeaders.PARTITION_KEY_RANGE_ID);
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/StoreClient.java:174]: perf impact?

any invocation to `HttpHeaders.getValue(name)` will invoke name.toLower
https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/core/azure-core/src/main/java/com/azure/core/http/HttpHeaders.java#L81

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosException.java [hunk 4]
```diff
@@ -54,7 +54,7 @@ protected CosmosException(int statusCode, String message, Map<String, String> re
         super(message, cause);
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosException.java:54]: Protected CTOR change will be another PR.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentServiceResponse.java [hunk 5]
```diff
@@ -17,32 +18,28 @@
...
         String[] headerNames = response.getResponseHeaderNames();
         String[] headerValues = response.getResponseHeaderValues();
 
-        this.headersMap = new HashMap<>(headerNames.length);
-
-        // Gets status code.
-        this.statusCode = response.getStatus();
-
         // Extracts headers.
+        this.headersMap = new HttpHeaders();
```
[the remark this answers is not in the document]
Comment (kirankumarkolli) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentServiceResponse.java:36]: Its possible, started a conversartion with core team. 
Default value of map is 16 and might suffice but need to perf test it.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/StoreClient.java (hunk 3 above)
Comment (kirankumarkolli) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/StoreClient.java:174]: Best way forward is to do perf testing and see the impact. 
And we can always follow-up on the optimization if the impact is with-in the bounds.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosException.java [hunk 6]
```diff
@@ -54,7 +54,7 @@ protected CosmosException(int statusCode, String message, Map<String, String> re
...
         this.statusCode = statusCode;
         this.requestTimeline = RequestTimeline.empty();
-        this.responseHeaders = responseHeaders == null ? new HashMap<>() : new HashMap<>(responseHeaders);
+        this.responseHeaders = responseHeaders == null ? new HttpHeaders() : new HttpHeaders(responseHeaders);
```
Comment (simplynaveen20) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosException.java:57]: Can we construct HttpHeaders when we are creating Exception and then pass it along ? , it may avoid one extra step of Map creation

Comment (kirankumarkolli) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosException.java:57]: Yup, that;s the next PR referred in the description.

**Comment by kirankumarkolli** [comment]
Here are the perf numbers

| counter   | RUN1 (master) | RUN2 (master) | RUN1(PR) | RUN2(PR)|
| ------------- |:-------------:| -----:|------:| --------:|
|         mean rate | 10332.04 | 10501.42 |10231.82| 9997.79 |
|     1-minute rate | 9848.01 | 10006.35 | 9663.18 | 9772.05 |
|     5-minute rate | 4979.89 | 4926.60 |  5041.40 | 4949.97 |
|    15-minute rate | 3647.43 | 3526.00 |  3757.96 | 3596.68 |

**Comment by kirankumarkolli** [comment]
With the latest PR 5 runs (including all optimizations)

| counter   | RUN1  | RUN2  | RUN3 | RUN4 | RUN5 |
| ------------- |:-------------:| -----:|------:| --------:| --------:|
| total time        | 92       |   93     |  88      | 91      | 95        |
|         mean rate | 10753.39 | 10695.11 | 11309.63 | 10954.95 | 10458.41 |
|     1-minute rate | 9987.19  | 10070.64 | 10271.09 | 10253.08 | 10185.69 |
|     5-minute rate | 5390.47  | 5407.82  | 5755.39  | 5562.00  | 5460.26  |
|    15-minute rate | 4148.75  | 4182.36  | 4555.01  | 4299.99  | 4195.58  |

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/StoreClient.java (hunk 3 above)
Comment (kirankumarkolli) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/StoreClient.java:163]: Check the perf numbers below.
